File: LeetCode/binarysearch/SplitArrayLargestSum.java
package binarysearch;

public class SplitArrayLargestSum {

    public int splitArray(int[] nums, int k) {
        int left = 0, right = 0;

        for (int num: nums) {
            left = Math.max(left, num);
            right += num;
        }

        while (left < right) {
            int mid = left + (right - left) / 2;

            if (countPartitions(nums, mid) <= k) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    private int countPartitions(int[] nums, int sum) {
        int total = 0;
        int partition = 0;

        for (int num: nums) {
            total += num;
            if (total > sum) {
                partition++;
                total = num;
            }
        }

        if (total > 0) {
            partition++;
        }

        return partition;
    }

    public static void main(String[] args) {
        SplitArrayLargestSum solution = new SplitArrayLargestSum();

        int[] nums = {1, 2, 3, 4, 5};
        int k = 2;
        System.out.println(solution.splitArray(nums, k)); // Output: 9
    }
}

File: LeetCode_Python/binarysearch/SplitArrayLargestSum.py
class SplitArrayLargestSum:
    def splitArray(self, nums: list[int], k: int) -> int:
        def calculate(target: int) -> int:
            total, count = 0, 0

            for n in nums:
                total += n
                if total > target:
                    count += 1
                    total = n

            if total > 0:
                count += 1

            return count

        left, right = max(nums), sum(nums)

        while left < right:
            mid = left + (right - left) // 2
            if calculate(mid) > k:
                left = mid + 1
            else:
                right = mid

        return left


# Time Complexity: O(n log(sum(nums))), where n is the length of the nums array.
# This is because we perform a binary search on the range from max(nums) to sum(nums),
# and for each mid value, we iterate through the nums array to calculate the number of subarrays.
# Space Complexity: O(1), as we are using a constant amount of space for variables.

if __name__ == "__main__":
    solution = SplitArrayLargestSum()

    print("Expected Output: 18")
    print("Actual Output:", solution.splitArray([7, 2, 5, 10, 8], 2))

    print("Expected Output: 9")
    print("Actual Output:", solution.splitArray([1, 2, 3, 4, 5], 2))

    print("Expected Output: 4")
    print("Actual Output:", solution.splitArray([1, 4, 4], 3))
